e2e/src/cmd.rs

```diff
@@ -22,3 +24,101 @@ pub fn for_each_batch(
 
     Ok(())
 }
+
+/// Runs a closure for each batch for the specified event history.
+pub fn for_each_batch<R, F>(
+    orderbook_file: impl AsRef<Path>,
+    mut report: R,
+    handler: F,
+) -> Result<R::Summary>
+where
+    R: Reporting,
+    F: Fn(&Sampler<R::Sample>, &ExchangeHistory, BatchId) -> Result<()> + Send + Sync,
+{
+    let history = ExchangeHistory::from_filestore(orderbook_file)?;
+
+    let batches = history.batches_until_now();
+    let mut progress = ProgressBar::on(io::stderr(), batches.batch_count());
+
+    let (samples_tx, samples_rx) = mpsc::channel();
+    let sampler = Sampler(samples_tx);
+
+    thread::scope(|scope| {
+        scope.spawn(move |_| {
+            if let Err(err) = batches
+                .collect::<Vec<_>>()
+                .into_par_iter()
+                .try_for_each_with(sampler.clone(), move |sampler, batch| {
+                    handler(&sampler, &history, batch)?;
+                    sampler.batch_complete()
+                })
+            {
+                sampler.error(err);
+            }
+        });
+
+        while let Ok(result) = samples_rx.recv() {
```

Comment:
This should probably be in a `scope.spawn` too because the intention is to have this run in parallel.

Reply:
So, the idea here was to have the main thread update the progress bar iterator and record the samples while a background thread processes the parallel . I chose this for a couple of reasons:
1. Not require another thread for this, we can just use the main thread and then join once we finish collecting
2. Not require `report` to also be `Send`